lib/internal/promise_hook.js

```diff
@@ -0,0 +1,97 @@
+'use strict';
+
+const {
+  ArrayPrototypeIndexOf,
+  ArrayPrototypeSplice,
+  ArrayPrototypePush,
+  FunctionPrototypeBind
+} = primordials;
+
+const { setPromiseHooks } = internalBinding('async_wrap');
+
+const hooks = {
+  init: [],
+  before: [],
+  after: [],
+  resolve: []
+}
+
+function initAll(promise, parent) {
+  for (const init of hooks.init) {
+    init(promise, parent)
+  }
+}
+
+function beforeAll(promise) {
+  for (const before of hooks.before) {
+    before(promise)
+  }
+}
+
+function afterAll(promise) {
+  for (const after of hooks.after) {
+    after(promise)
+  }
+}
+
+function resolveAll(promise) {
+  for (const resolve of hooks.resolve) {
+    resolve(promise)
+  }
+}
+
+function maybeFastPath(list, runAll) {
+  return list.length > 1 ? runAll : list[0];
+}
+
+function update() {
+  const init = maybeFastPath(hooks.init, initAll);
+  const before = maybeFastPath(hooks.before, beforeAll);
+  const after = maybeFastPath(hooks.after, afterAll);
+  const resolve = maybeFastPath(hooks.resolve, resolveAll);
+  setPromiseHooks(init, before, after, resolve);
+}
+
+function stop(list, hook) {
+  const index = ArrayPrototypeIndexOf(list, hook);
+  if (index >= 0) {
+    ArrayPrototypeSplice(list, index, 1);
+    update();
+  }
+}
+
+function makeUseHook(list) {
+  return (hook) => {
+    ArrayPrototypePush(list, hook);
+    update();
+    return FunctionPrototypeBind(stop, null, list, hook);
+  }
+}
+
+const onInit = makeUseHook(hooks.init);
+const onBefore = makeUseHook(hooks.before);
+const onAfter = makeUseHook(hooks.after);
+const onResolve = makeUseHook(hooks.resolve);
+
+function createHook({ init, before, after, resolve } = {}) {
+  const hooks = [];
+
+  if (init) ArrayPrototypePush(hooks, onInit(init));
+  if (before) ArrayPrototypePush(hooks, onBefore(before));
+  if (after) ArrayPrototypePush(hooks, onAfter(after));
+  if (resolve) ArrayPrototypePush(hooks, onResolve(resolve));
+
+  return () => {
```

Comment:
I see it as API fragmentation to deviate from the `enable` and `disable` methods in `async_hooks`.

Reply:
What do you think of `createHook` returning an object with `enable`/`disable` while the `on*` functions continue to just return `stop` functions?

To me, the API of `async_hooks` is more complicated than it needs to be, especially here where a lot of users are likely only going to care about the `init` event. I'm not a fan of explicitly needing to call `enable` at the start--why even _create_ the hooks if you aren't ready to enable them yet? Personally I'd rather just supply a function again if I want to attach it again rather that wrapping it in an unnecessary tracking object that does nothing else except attach and detach it. But then I also tend to prefer more functional code in general. 🤷

Reply:
Sometimes it is useful to be able to have a state in its own scope, like a `WeekMap` for the promises, while also delaying tracking until an event, for example `server.on('listen')`.

I don't think I had a strong opinion on what the default was (enabled or disabled), but there is a real benefit in being able to enable hooks again after they have been disabled.

Reply:
Sure, but I don't see why it needs an object to manage enabling it again when you could just pass the function in again. 🤷‍♂️

Reply:
The scope where you enable again may not be the scope of the WeekMap. In general, I don't think hooks should be functional. Hooks don't return anything and inherently depend on mutating a state to do something, so they will never truly be functional.

Reply:
Not sure what you are referring to about a WeakMap. As for if it's functional or not, I don't think that matters here. I just find it simpler to use functions rather than objects. It's also safer as internals can’t be tampered with. If people feel strongly that it should be object-based though, I'm happy to reconsider.

Reply:
Thinking about it more, I would say making it look more like async_hooks would actually be very confusing. More than just the specific intent being different, the actual functionality is different enough that it would just be a source of confusion, especially if we opt to move the API back into the async_hooks module. The set of lifecycle events does not match what async_hooks produces as there's no destroy event, and the hook functions all receive the promise object, there is no ID mechanism like async_hooks has, so the hook functions are different enough the making the API look too similar to async_hooks would most likely just confuse people that would expect functionality to be identical when it's not.

Reply:
I'm torn on the enable/disable question. I *prefer* the function return but, with the enable/disable pattern, I can create the hook but more precisely control when it is enabled. I like having that ability to separate the creation of the hook from the enabling of it. I can live with this API design tho.